### ISSUE TITLE
Noob Question : Access to default context and NSFetchedResultsController

### ISSUE THREAD

**Issue by wm-j-ray** [JohnEstropia/CoreStore#76]
John,

Just getting started with CoreStore, so...

If I 'needed' to use another complementary framework that needs a NSFetchedResultsController and a context for reading//writing, what is the safest way to create/access them without breaking a CoreStore dataStack set up like so:

    struct DataStoreService {
    
        static let coreDataStack: DataStack = {
        
            let dataStack = DataStack(modelName: "AppName")
        
            do {
                try dataStack.addSQLiteStoreAndWait(
                    fileName: "AppName.sqlite",
                    configuration:nil,
                    resetStoreOnModelMismatch: true
                )
            
            } catch {
                print("Failed creating dataStack with error: \(error as NSError)")
            }
            return dataStack
        }()
    
    }

And the stack is used from an interactor like structure using something like this:

      func fetchAllWidgets() -> [WidgetsMO]
      {
        return DataStoreService.coreDataStack.fetchAll(From(WidgetsMO))!
       }

I 'think' what I'm looking for is just a way to access the correct context to used for the NSFetchedResultsController (yours ideally, so I can use CoreStore fetching and queries) and the best way to deal with CoreStore transactions.  If I can't access your NSFetchedController I would create one. Monitors and Observers, with my level of knowledge about CoreStore, seem a little beyond me.

And what are the pitfalls and downside to doing something like this?  

The library is https://github.com/DenHeadless/DTTableViewManager by the way (which lightens up tableViews by using a manager (https://github.com/DenHeadless/DTModelStorage) to bind the model against tableViewCells), and I envision using it primarily for reads and using the CoreStore functionality for hopefully all of the writes.  

Basically, I would really like to somehow wire CoreStore and DTTableViewManager together.

I want to make sure I'm aware of all of the potential issues and pick the best approach before I go any further.

Thanx in Advance

**Comment by JohnEstropia** [comment]
The only NSManagedObjectContext exposed by CoreStore is the context held by `UnsafeDataTransaction`s.
```swift
let transaction = CoreStore.beginUnsafe() // store this somewhere so it doesn't get deallocated
let context = transaction.internalContext // this is what you need and you're free to do dirty things to it :)
```
See the documentation on `UnsafeDataTransaction.internalContext`: https://github.com/JohnEstropia/CoreStore/blob/master/CoreStore/Saving%20and%20Processing/UnsafeDataTransaction.swift#L153

**Comment by JohnEstropia** [comment]
Also, note that this means the objects you are working on are not from the main context. Don't mix objects fetched from the `DataStack` to the objects
```swift
// BAD!!
let someObject = CoreStore.fetchOne(From(...))

let transaction = CoreStore.beginUnsafe() // store this somewhere so it doesn't get deallocated
let context = transaction.internalContext
let anotherObject = // fetch from transaction/context
anotherObject.someRelationship = someObject // BAD!
```

**Comment by JohnEstropia** [comment]
If all you need is an `NSFetchedResultsController` instance, CoreStore can also create one for you. See 
[NSFetchedResultsController+Convenience.swift](https://github.com/JohnEstropia/CoreStore/blob/master/CoreStore/Convenience%20Helpers/NSFetchedResultsController%2BConvenience.swift#L37)

**Comment by wm-j-ray** [comment]
Thanks for the help.  Let me make sure I understand...

So I can call NSFetchedResultsController.createForStack I would be using the correct context.  I could use that NSFetchedResultsController to drive the tableView.  Can you give me an example of what that call would look like using my dataStack? (lame, I know, but thanks in advance)

And, if I update I wouldn't be able to use the default CoreStore methods even though I'm in the proper context but rather I would have to do a save() manually or an UnsafeDataTransaction commit.

**Comment by JohnEstropia** [comment]
Let me push an update so the API call is simpler. Once it's up you should be able to call with 
```swift
let controller = NSFetchedResultsController.createForStack(
    CoreStore.defaultStack,
    From(MyEntity),
    SectionBy("sectionKey"), // optional
    Where("value > %@", minValue), // optional
    OrderBy(.Ascending("value")) // optional
)
```

If I understand how DTTableViewManager works, it's only for displaying data, correct? If so then the NSFetchedResultsController created with CoreStore should notify DTTableViewManager without any issues, and you can still make updates using any of CoreStore's transactions (not just UnsafeDataTransactions).

**Comment by JohnEstropia** [comment]
Check out version 1.6.8. Renamed `createForStack()` to `createFor()` and now accepts either `DataStack` or `UnsafeDataTransaction`

**Comment by wm-j-ray** [comment]
Excellent.  Thanks so much.
On to the next.

**Comment by wm-j-ray** [comment]
John,

I pulled down 1.6.8 and I must be an idiot because I can't get the stub you provided to work:

    let controller = NSFetchedResultsController.createForStack(
    CoreStore.defaultStack,
    From(MyEntity),
    SectionBy("sectionKey"), // optional
    Where("value > %@", minValue), // optional
    OrderBy(.Ascending("value")) // optional
)

How do the optional parameters work?  Why can't I do this?

	let controller = NSFetchedResultsController.createForStack(
		CoreStore.defaultStack,
		From(NoteMO))

I get a missing fetchClauses in call error.  

I've also tried this:

    let fetchreq = NSFetchRequest(entityName: "NoteMO")
    let controller = NSFetchedResultsController.createForStack(CoreStore.defaultStack, fetchRequest: fetchreq, fetchClauses: nil)

I'm not sure what I'm doing incorrectly.

**Comment by JohnEstropia** [comment]
Sorry for the confusion, I renamed `createForStack()` to `createFor()`. The current `createForStack()` is deprecated and requires argument labels. Here's the correct syntax
```swift
let controller = NSFetchedResultsController.createFor(
    CoreStore.defaultStack,
    From(MyEntity),
    SectionBy("sectionKey"), // optional
    Where("value > %@", minValue), // optional
    OrderBy(.Ascending("value")) // optional
)
```

**Comment by JohnEstropia** [comment]
By the way,
```swift
let controller = NSFetchedResultsController.createFor(
    CoreStore.defaultStack,
    From(NoteMO))
```
will compile, but will raise an assertion because NSFetchedResultsControllers requires sorting information. Provide an `OrderBy` clause.

I'll update the documentation.

**Comment by wm-j-ray** [comment]
Got it and many thanks.
I'll give it a shot.

**Comment by wm-j-ray** [comment]
OK, now I'm perplexed (and probably don't know what I'm doing)

1.  Set up the stack like so:

	    func startService() {

		    CoreStore.defaultStack = DataStack(
			    modelName: "WriteTrack"
			    //migrationChain: ["MyStore", "MyStoreV2", "MyStoreV3"]
		    )
        
            do {
                try CoreStore.addSQLiteStoreAndWait(
                    fileName: "WriteTrack.sqlite",
                    configuration:nil,
                    resetStoreOnModelMismatch: true
                )
            
            } catch {
                print("Failed creating CoreStore with error: \(error as NSError)")
            }

        }

2.  Fetch all records normal CoreStore way, like so:

        func fetchAllNotes() -> [NoteMO]
        {
            return CoreStore.fetchAll(From(NoteMO))! 
        }

Returns all of the records.


3.  Try with a NSFetchedController like so:

    	func test() {
		let controller = NSFetchedResultsController.createFor(
			CoreStore.defaultStack,
			From(NoteMO), OrderBy(.Ascending("content")))
		
		do {
			try controller.performFetch()
		} catch {
			print("Failed performing fetch with error: \(error as NSError)")
		}
		controller.fetchedObjects
		// breakpoint here
    }


And I see the controller, but there is nothing in it...

![screen shot 2016-05-27 at 9 52 28 pm](https://cloud.githubusercontent.com/assets/770852/15624418/771cc112-2455-11e6-990c-b8cae7073879.png)

**Comment by JohnEstropia** [comment]
Did you set the NSFetchedResultsControllerDelegate? NSFetchedResultsControllers are lazy, and will not do actual fetching until you access it's array or unless you have a delegate that implements `controllerDidChangeContent(_:)`

**Comment by JohnEstropia** [comment]
Try to print the content after performFetch()
```
try controller.performFetch()
print(controller.fetchedObjects)
```

**Comment by wm-j-ray** [comment]
That was it. Thanks again and I can't tell you how much I appreciate your patience.

**Comment by JohnEstropia** [comment]
No problem! Feel free to ask anything else anytime :)
